File: javascript/StudentGrades.js
const readline = require("readline");

let subjects = [];
let studentNames = [];
let marks = [];

function askFactory() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const ask = (question) =>
    new Promise((resolve) => {
      rl.question(question, (answer) => resolve(answer));
    });

  return { rl, ask };
}

async function registerSubjects(ask) {
  console.log("\nRegister subjects:");
  for (let i = 0; i < subjects.length; i += 1) {
    subjects[i] = await ask(`Enter subject ${i + 1} name: `);
  }
}

async function registerStudentsAndMarks(ask, numStudents, numSubjects) {
  for (let i = 0; i < numStudents; i += 1) {
    studentNames[i] = await ask(`\nEnter student ${i + 1} name: `);

    for (let j = 0; j < numSubjects; j += 1) {
      marks[i][j] = Number(await ask(`Enter marks for ${subjects[j]}: `));
    }
  }
}

function calculateAverage(studentMarks, numSubjects) {
  let total = 0;
  for (const mark of studentMarks) {
    total += mark;
  }
  return total / numSubjects;
}

function determineGrade(avgMarks) {
  if (avgMarks >= 90) return "A";
  if (avgMarks >= 80) return "B";
  if (avgMarks >= 70) return "C";
  if (avgMarks >= 60) return "D";
  return "F";
}

function displayReport(numStudents, numSubjects) {
  console.log("\nStudent Report:");
  console.log("------------------------------------");
  console.log(`${"Student Name".padEnd(15)} ${"Avg Marks".padEnd(10)} ${"Grade".padEnd(10)}`);
  console.log("------------------------------------");

  for (let i = 0; i < numStudents; i += 1) {
    const avgMarks = calculateAverage(marks[i], numSubjects);
    const grade = determineGrade(avgMarks);
    console.log(`${studentNames[i].padEnd(15)} ${avgMarks.toFixed(2).padEnd(10)} ${grade.padEnd(10)}`);
  }
}

async function main() {
  const { rl, ask } = askFactory();

  const numStudents = Number(await ask("Enter the number of students: "));
  const numSubjects = Number(await ask("Enter the number of subjects: "));

  subjects = Array(numSubjects).fill("");
  studentNames = Array(numStudents).fill("");
  marks = Array.from({ length: numStudents }, () => Array(numSubjects).fill(0));

  await registerSubjects(ask);
  await registerStudentsAndMarks(ask, numStudents, numSubjects);
  displayReport(numStudents, numSubjects);

  rl.close();
}

main();

File: python/student_grades.py
subjects = []
student_names = []
marks = []


def register_subjects():
    print("\nRegister subjects:")
    for i in range(len(subjects)):
        subjects[i] = input(f"Enter subject {i + 1} name: ")


def register_students_and_marks(num_students, num_subjects):
    for i in range(num_students):
        student_names[i] = input(f"\nEnter student {i + 1} name: ")

        for j in range(num_subjects):
            marks[i][j] = int(input(f"Enter marks for {subjects[j]}: "))


def calculate_average(student_marks, num_subjects):
    total = 0
    for mark in student_marks:
        total += mark
    return total / num_subjects


def determine_grade(avg_marks):
    if avg_marks >= 90:
        return "A"
    if avg_marks >= 80:
        return "B"
    if avg_marks >= 70:
        return "C"
    if avg_marks >= 60:
        return "D"
    return "F"


def display_report(num_students, num_subjects):
    print("\nStudent Report:")
    print("------------------------------------")
    print(f"{'Student Name':<15} {'Avg Marks':<10} {'Grade':<10}")
    print("------------------------------------")

    for i in range(num_students):
        avg_marks = calculate_average(marks[i], num_subjects)
        grade = determine_grade(avg_marks)
        print(f"{student_names[i]:<15} {avg_marks:<10.2f} {grade:<10}")


def main():
    num_students = int(input("Enter the number of students: "))
    num_subjects = int(input("Enter the number of subjects: "))

    global subjects, student_names, marks
    subjects = ["" for _ in range(num_subjects)]
    student_names = ["" for _ in range(num_students)]
    marks = [[0 for _ in range(num_subjects)] for _ in range(num_students)]

    register_subjects()
    register_students_and_marks(num_students, num_subjects)
    display_report(num_students, num_subjects)


if __name__ == "__main__":
    main()
